src/symbols.ts

```diff
@@ -0,0 +1,4 @@
+import { InjectionKey } from "vue";
+import { TabData } from './types';
+
+export const TabsKey: InjectionKey<Record<string,TabData>> = Symbol( 'Tabs' );
```

Comment:
oof...what's the advantage of defining this string in this way? or is it something TypeScript required you to do?

Reply:
I agree that this is getting pretty arcane (generics wrapping generics).

#### Short explanation
Normally provide/inject can pass anything (which TS doesn't like). By using the InjectionKey generic type, we can ensure that a given use of provide/inject will always have a specific type.

#### Longer explanation
Using `InjectionKey` gives you type-safety when using provide & inject (which otherwise can return anything).

Normally you use provide/inject with a string, and it can be used to pass any sort of value at all. With InjectionKey, we are saying: this provided or injected value will always be of the following type (that's why the type is `InjectionKey<T>` – the type contains a placeholder for another type we will provide; this is a [generic](https://www.typescriptlang.org/docs/handbook/2/generics.html) in TS).

So `const TabsKey: InjectionKey<Record<string,TabData>> = Symbol( 'Tabs' );` is equivalent to saying: this key can only be used to pass a Record value with string keys and values of type `TabData`. 

If `provide( TabsKey, tabsData );` ever passes in data that no longer matches this type, Typescript will complain.
On the receiving side, Typescript already knows that `const tabsData = inject( TabsKey )` will have a type that matches what is specified in that symbols file (plus `undefined`, since it's possible that no value was sent).

This blog post was my main source of reference as I was figuring this out: https://logaretm.com/blog/2020-12-23-type-safe-provide-inject

Reply:
If you didn't use InjectionKey but were still using Typescript, this code would give you a big headache:

```ts
const tabsData = inject( TabsKey );
```

`tabsData` would have type `any` as far as TS was concerned, and we would be punished for that.

You could do this:

```ts
const tabsData = inject( TabsKey ) as Record<string,TabData>;
```

Which would shut up the compiler, but it's kind of a false sense of security – our code will expect `tabsData` to be the type we've specified but we have not actually done anything to ensure that this is the only value that can be provided.

In general it sounds like using direct type assertion is something that should be avoided whenever possible.

Reply:
Thanks for the helpful explanation! This is another example of TypeScript doing something incredibly valuable in an unfortunately confusing way. This sort of explanation would be a good candidate for on-wiki documentation on using TypeScript + Vue without wanting to scratch your eyes out (which is the working title of the article I intend to create...)